src/garage/torch/policies/context_conditioned_policy.py

```diff
@@ -0,0 +1,271 @@
+        else:
+            self._context = torch.cat([self._context, data], dim=1)
+
+    def infer_posterior(self, context):
+        """Compute q(z|c) as a function of input context and sample new z.
+
+        Args:
+            context (torch.Tensor): Context values.
+
+        """
+        params = self._context_encoder(context)
+        params = params.view(context.size(0), -1,
+                             self._context_encoder._output_dim)
+        # with probabilistic z, predict mean and variance of q(z | c)
+        if self._use_ib:
+            mu = params[..., :self._latent_dim]
+            sigma_squared = F.softplus(params[..., self._latent_dim:])
+            z_params = [
+                _product_of_gaussians(m, s)
+                for m, s in zip(torch.unbind(mu), torch.unbind(sigma_squared))
+            ]
+            self.z_means = torch.stack([p[0] for p in z_params])
+            self.z_vars = torch.stack([p[1] for p in z_params])
+        else:
+            self.z_means = torch.mean(params, dim=1)
+        self.sample_from_belief()
+
+    # pylint: disable=arguments-differ
+    def forward(self, obs, context):
+        """Given observations and context, get actions and probs from policy.
+
+        Args:
+            obs (torch.Tensor): Observation values.
+            context (torch.Tensor): Context values.
+
+        Returns:
+            torch.Tensor: Output action values.
+            torch.Tensor: z values.
+
+        """
+        self.infer_posterior(context)
+        self.sample_from_belief()
+        task_z = self.z
+
+        # task, batch
+        t, b, _ = obs.size()
+        obs = obs.view(t * b, -1)
+        task_z = [z.repeat(b, 1) for z in task_z]
+        task_z = torch.cat(task_z, dim=0)
+
+        # run policy, get log probs and new actions
+        obs_z = torch.cat([obs, task_z.detach()], dim=1)
+        dist = self._policy(obs_z)
+        pre_tanh, actions = dist.base_dist.rsample_with_pre_tanh_value()
+        log_pi = dist.base_dist.log_prob(value=actions,
+                                         pre_tanh_value=pre_tanh)
+        log_pi = log_pi.unsqueeze(1)
+        mean = dist.mean.to('cpu').detach().numpy()
+        log_std = (dist.variance**.5).log().to('cpu').detach().numpy()
+
+        return (actions, mean, log_std, log_pi, pre_tanh), task_z
+
+    def get_action(self, obs):
+        """Sample action from the policy, conditioned on the task embedding.
+
+        Args:
+            obs (torch.Tensor): Observation values.
+
+        Returns:
+            torch.Tensor: Output action values.
+
+        """
+        z = self.z
+        obs = tu.from_numpy(obs[None])
+        obs_in = torch.cat([obs, z], dim=1)
+        action, info = self._policy.get_action(obs_in)
+        action = np.squeeze(action, axis=0)
+        info['mean'] = np.squeeze(info['mean'], axis=0)
+        return action, info
+
+    def compute_kl_div(self):
+        """Compute KL( q(z|c) || p(z) ).
```

Comment:
you can use a math environment to show these very nicely in LaTeX.

see https://github.com/rlworkgroup/garage/blob/master/src/garage/_dtypes.py for an example